File: csharp-selenium-nunit-reqnroll/SauceDemo.UiTests/Pages/CheckoutInformationPage.cs
using OpenQA.Selenium;
using SauceDemo.UiTests.TestData;

namespace SauceDemo.UiTests.Pages;

public sealed class CheckoutInformationPage : BasePage
{
    private static readonly By FirstNameInput = By.CssSelector("[data-test='firstName']");
    private static readonly By LastNameInput = By.CssSelector("[data-test='lastName']");
    private static readonly By PostalCodeInput = By.CssSelector("[data-test='postalCode']");
    private static readonly By ContinueButton = By.CssSelector("[data-test='continue']");
    private static readonly By ErrorBanner = By.CssSelector("[data-test='error']");

    public CheckoutInformationPage(IWebDriver driver) : base(driver)
    {
    }

    public CheckoutOverviewPage FillAndContinue(CheckoutInfo info)
    {
        Type(FirstNameInput, info.FirstName);
        Type(LastNameInput, info.LastName);
        Type(PostalCodeInput, info.PostalCode);
        Click(ContinueButton);
        return new CheckoutOverviewPage(Driver);
    }

    /// <summary>Leaves the postal code blank and submits, for the "postal code is required" negative scenario.</summary>
    public void SubmitWithoutPostalCode(CheckoutInfo info)
    {
        Type(FirstNameInput, info.FirstName);
        Type(LastNameInput, info.LastName);
        Click(ContinueButton);
    }

    public string GetErrorMessage() => TextOf(ErrorBanner);

    public bool HasError() => IsVisible(ErrorBanner);
}

File: ts-playwright-cucumber/src/pages/checkoutInformationPage.ts
import type { Page } from 'playwright';
import { BasePage } from './basePage.ts';
import { CheckoutOverviewPage } from './checkoutOverviewPage.ts';
import type { CheckoutInfo } from '../testdata/checkoutInfo.ts';

const FIRST_NAME_INPUT = "[data-test='firstName']";
const LAST_NAME_INPUT = "[data-test='lastName']";
const POSTAL_CODE_INPUT = "[data-test='postalCode']";
const CONTINUE_BUTTON = "[data-test='continue']";
const ERROR_BANNER = "[data-test='error']";

export class CheckoutInformationPage extends BasePage {
  constructor(page: Page) {
    super(page);
  }

  async fillAndContinue(info: CheckoutInfo): Promise<CheckoutOverviewPage> {
    await this.type(FIRST_NAME_INPUT, info.firstName);
    await this.type(LAST_NAME_INPUT, info.lastName);
    await this.type(POSTAL_CODE_INPUT, info.postalCode);
    await this.click(CONTINUE_BUTTON);
    return new CheckoutOverviewPage(this.page);
  }

  /** Leaves the postal code blank and submits, for the "postal code is required" negative scenario. */
  async submitWithoutPostalCode(info: CheckoutInfo): Promise<void> {
    await this.type(FIRST_NAME_INPUT, info.firstName);
    await this.type(LAST_NAME_INPUT, info.lastName);
    await this.click(CONTINUE_BUTTON);
  }

  getErrorMessage(): Promise<string> {
    return this.textOf(ERROR_BANNER);
  }

  hasError(): Promise<boolean> {
    return this.waitAndCheckVisible(ERROR_BANNER);
  }
}
